File: android/src/main/kotlin/dev/flutterberlin/flutter_gemma/VectorStore.kt
package dev.flutterberlin.flutter_gemma

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import kotlin.math.sqrt

class VectorStore(
    private val context: Context,
    private val dimension: Int? = null  // null = auto-detect from first document
) {
    private var dbHelper: VectorDatabaseHelper? = null
    private var database: SQLiteDatabase? = null
    private var detectedDimension: Int? = null

    companion object {
        const val DATABASE_NAME = "flutter_gemma_vectors.db"
        const val DATABASE_VERSION = 2  // Increment for schema change

        // Common dimensions (informational only)
        const val DIM_GECKO_SMALL = 256
        const val DIM_MINI_LM = 384
        const val DIM_BERT_BASE = 768
        const val DIM_BERT_LARGE = 1024
        const val DIM_COHERE_V3 = 1024
        const val DIM_OPENAI_ADA = 1536
        const val DIM_OPENAI_LARGE = 3072
        const val DIM_QWEN_3 = 4096
        
        const val TABLE_DOCUMENTS = "documents"
        const val COLUMN_ID = "id"
        const val COLUMN_CONTENT = "content"
        const val COLUMN_EMBEDDING = "embedding"
        const val COLUMN_METADATA = "metadata"
        const val COLUMN_CREATED_AT = "created_at"
    }
    
    private inner class VectorDatabaseHelper(context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
        override fun onCreate(db: SQLiteDatabase) {
            val createTableSQL = """
                CREATE TABLE $TABLE_DOCUMENTS (
                    $COLUMN_ID TEXT PRIMARY KEY,
                    $COLUMN_CONTENT TEXT NOT NULL,
                    $COLUMN_EMBEDDING BLOB NOT NULL,
                    $COLUMN_METADATA TEXT,
                    $COLUMN_CREATED_AT INTEGER DEFAULT (strftime('%s', 'now'))
                );
                CREATE INDEX idx_created_at ON $TABLE_DOCUMENTS($COLUMN_CREATED_AT);
            """.trimIndent()
            db.execSQL(createTableSQL)
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            // Simple DROP is acceptable: we changed embedding storage format (JSON→BLOB)
            // and migration is not possible. Users will need to re-index documents.
            db.execSQL("DROP TABLE IF EXISTS $TABLE_DOCUMENTS")
            onCreate(db)
        }
    }
    
    fun initialize(databasePath: String) {
        dbHelper = VectorDatabaseHelper(context)
        database = dbHelper?.writableDatabase
    }
    
    fun addDocument(id: String, content: String, embedding: List<Double>, metadata: String?) {
        val db = database ?: throw IllegalStateException("Database not initialized")

        // Auto-detect dimension from first document
        if (detectedDimension == null) {
            detectedDimension = dimension ?: embedding.size

            // Validate if dimension was specified
            if (dimension != null && dimension != embedding.size) {
                throw IllegalArgumentException(
                    "Embedding dimension mismatch: expected $dimension, got ${embedding.size}"
                )
            }
        }

        // Validate dimension consistency for subsequent documents
        if (embedding.size != detectedDimension) {
            throw IllegalArgumentException(
                "Embedding dimension mismatch: expected $detectedDimension, got ${embedding.size}"
            )
        }

        // Convert to binary BLOB
        val embeddingBlob = embeddingToBlob(embedding)

        val values = ContentValues().apply {
            put(COLUMN_ID, id)
            put(COLUMN_CONTENT, content)
            put(COLUMN_EMBEDDING, embeddingBlob)
            put(COLUMN_METADATA, metadata)
        }

        db.insertWithOnConflict(TABLE_DOCUMENTS, null, values, SQLiteDatabase.CONFLICT_REPLACE)
    }
    
    fun searchSimilar(queryEmbedding: List<Double>, topK: Int, threshold: Double): List<RetrievalResult> {
        val db = database ?: throw IllegalStateException("Database not initialized")

        // Validate query embedding dimension
        if (detectedDimension != null && queryEmbedding.size != detectedDimension) {
            throw IllegalArgumentException(
                "Query embedding dimension mismatch: expected $detectedDimension, got ${queryEmbedding.size}"
            )
        }

        val cursor = db.query(TABLE_DOCUMENTS, null, null, null, null, null, null)
        val results = mutableListOf<Pair<RetrievalResult, Double>>()

        cursor.use {
            val idIndex = cursor.getColumnIndexOrThrow(COLUMN_ID)
            val contentIndex = cursor.getColumnIndexOrThrow(COLUMN_CONTENT)
            val embeddingIndex = cursor.getColumnIndexOrThrow(COLUMN_EMBEDDING)
            val metadataIndex = cursor.getColumnIndexOrThrow(COLUMN_METADATA)

            while (cursor.moveToNext()) {
                val id = cursor.getString(idIndex)
                val content = cursor.getString(contentIndex)
                val embeddingBlob = cursor.getBlob(embeddingIndex)
                val metadata = cursor.getString(metadataIndex)

                // Convert BLOB to embedding
                val embedding = blobToEmbedding(embeddingBlob)

                val similarity = cosineSimilarity(queryEmbedding, embedding)

                if (similarity >= threshold) {
                    val result = RetrievalResult(
                        id = id,
                        content = content,
                        similarity = similarity,
                        metadata = metadata
                    )
                    results.add(result to similarity)
                }
            }
        }

        return results
            .sortedByDescending { it.second }
            .take(topK)
            .map { it.first }
    }
    
    fun getStats(): VectorStoreStats {
        val db = database ?: throw IllegalStateException("Database not initialized")

        val cursor = db.rawQuery("SELECT COUNT(*) FROM $TABLE_DOCUMENTS", null)
        val count = cursor.use {
            if (it.moveToFirst()) it.getLong(0) else 0L
        }

        return VectorStoreStats(
            documentCount = count,
            vectorDimension = (detectedDimension ?: 0).toLong()
        )
    }
    
    fun clear() {
        val db = database ?: throw IllegalStateException("Database not initialized")
        db.delete(TABLE_DOCUMENTS, null, null)
    }
    
    private fun cosineSimilarity(a: List<Double>, b: List<Double>): Double {
        if (a.size != b.size) return 0.0

        var dotProduct = 0.0
        var normA = 0.0
        var normB = 0.0

        for (i in a.indices) {
            dotProduct += a[i] * b[i]
            normA += a[i] * a[i]
            normB += b[i] * b[i]
        }

        return if (normA != 0.0 && normB != 0.0) {
            dotProduct / (sqrt(normA) * sqrt(normB))
        } else 0.0
    }

    /**
     * Convert embedding List<Double> to binary BLOB (float32)
     *
     * Format: Little-endian float32 array
     * Size: dimension * 4 bytes (e.g., 768D = 3,072 bytes)
     */
    private fun embeddingToBlob(embedding: List<Double>): ByteArray {
        val buffer = java.nio.ByteBuffer.allocate(embedding.size * 4)
        buffer.order(java.nio.ByteOrder.LITTLE_ENDIAN)
        embedding.forEach { buffer.putFloat(it.toFloat()) }
        return buffer.array()
    }

    /**
     * Convert binary BLOB (float32) to embedding List<Double>
     */
    private fun blobToEmbedding(blob: ByteArray): List<Double> {
        val buffer = java.nio.ByteBuffer.wrap(blob)
        buffer.order(java.nio.ByteOrder.LITTLE_ENDIAN)
        return (0 until blob.size / 4).map {
            buffer.getFloat(it * 4).toDouble()
        }
    }
}

File: ios/Classes/VectorStore.swift
import Foundation
import SQLite3

/// iOS VectorStore implementation - equivalent to Android VectorStore.kt
/// Stores document embeddings in SQLite with binary BLOB format
class VectorStore {

    // MARK: - Properties

    private var db: OpaquePointer?
    private var dimension: Int?
    private var detectedDimension: Int?

    // Database schema
    private static let databaseVersion = 2
    private static let tableName = "documents"
    private static let columnId = "id"
    private static let columnContent = "content"
    private static let columnEmbedding = "embedding"
    private static let columnMetadata = "metadata"
    private static let columnCreatedAt = "created_at"

    // Common dimensions (informational only)
    static let DIM_GECKO_SMALL = 256
    static let DIM_MINI_LM = 384
    static let DIM_BERT_BASE = 768
    static let DIM_BERT_LARGE = 1024
    static let DIM_COHERE_V3 = 1024
    static let DIM_OPENAI_ADA = 1536
    static let DIM_OPENAI_LARGE = 3072
    static let DIM_QWEN_3 = 4096

    // MARK: - Initialization

    /// Initialize VectorStore with optional dimension
    /// - Parameter dimension: Expected embedding dimension (nil = auto-detect)
    init(dimension: Int? = nil) {
        self.dimension = dimension
    }

    /// Initialize database at specified path
    func initialize(databasePath: String) throws {
        // Close existing connection if any
        close()

        // Open SQLite database
        if sqlite3_open(databasePath, &db) != SQLITE_OK {
            throw VectorStoreError.databaseOpenFailed("Failed to open database at: \(databasePath)")
        }

        // Create table if not exists
        try createTable()
    }

    // MARK: - Public Methods

    /// Add document with embedding to vector store
    func addDocument(id: String, content: String, embedding: [Double], metadata: String?) throws {
        guard let db = db else {
            throw VectorStoreError.databaseNotInitialized
        }

        // Auto-detect dimension from first document
        if detectedDimension == nil {
            detectedDimension = dimension ?? embedding.count

            // Validate if dimension was specified
            if let expectedDim = dimension, expectedDim != embedding.count {
                throw VectorStoreError.dimensionMismatch(
                    expected: expectedDim,
                    actual: embedding.count
                )
            }
        }

        // Validate dimension consistency
        if embedding.count != detectedDimension {
            throw VectorStoreError.dimensionMismatch(
                expected: detectedDimension!,
                actual: embedding.count
            )
        }

        // Convert embedding to binary BLOB
        let embeddingBlob = embeddingToBlob(embedding)

        // Prepare INSERT statement
        let insertSQL = """
        INSERT OR REPLACE INTO \(Self.tableName)
        (\(Self.columnId), \(Self.columnContent), \(Self.columnEmbedding), \(Self.columnMetadata))
        VALUES (?, ?, ?, ?);
        """

        var stmt: OpaquePointer?

        if sqlite3_prepare_v2(db, insertSQL, -1, &stmt, nil) == SQLITE_OK {
            sqlite3_bind_text(stmt, 1, (id as NSString).utf8String, -1, nil)
            sqlite3_bind_text(stmt, 2, (content as NSString).utf8String, -1, nil)

            // Bind BLOB
            embeddingBlob.withUnsafeBytes { ptr in
                sqlite3_bind_blob(stmt, 3, ptr.baseAddress, Int32(embeddingBlob.count), nil)
            }

            if let metadata = metadata {
                sqlite3_bind_text(stmt, 4, (metadata as NSString).utf8String, -1, nil)
            } else {
                sqlite3_bind_null(stmt, 4)
            }

            if sqlite3_step(stmt) != SQLITE_DONE {
                sqlite3_finalize(stmt)
                throw VectorStoreError.insertFailed("Failed to insert document")
            }
        } else {
            throw VectorStoreError.insertFailed("Failed to prepare insert statement")
        }

        sqlite3_finalize(stmt)
    }

    /// Search for similar documents using cosine similarity
    func searchSimilar(
        queryEmbedding: [Double],
        topK: Int,
        threshold: Double
    ) throws -> [RetrievalResult] {
        guard let db = db else {
            throw VectorStoreError.databaseNotInitialized
        }

        // Validate query embedding dimension
        if let detectedDim = detectedDimension, queryEmbedding.count != detectedDim {
            throw VectorStoreError.dimensionMismatch(
                expected: detectedDim,
                actual: queryEmbedding.count
            )
        }

        let querySQL = """
        SELECT \(Self.columnId), \(Self.columnContent), \(Self.columnEmbedding), \(Self.columnMetadata)
        FROM \(Self.tableName);
        """

        var stmt: OpaquePointer?
        var results: [(result: RetrievalResult, similarity: Double)] = []

        if sqlite3_prepare_v2(db, querySQL, -1, &stmt, nil) == SQLITE_OK {
            while sqlite3_step(stmt) == SQLITE_ROW {
                // Extract columns
                let id = String(cString: sqlite3_column_text(stmt, 0))
                let content = String(cString: sqlite3_column_text(stmt, 1))

                // Extract BLOB
                if let embeddingBlob = sqlite3_column_blob(stmt, 2) {
                    let embeddingSize = sqlite3_column_bytes(stmt, 2)
                    let embeddingData = Data(bytes: embeddingBlob, count: Int(embeddingSize))
                    let docEmbedding = blobToEmbedding(embeddingData)

                    // Calculate similarity
                    let similarity = VectorUtils.cosineSimilarity(queryEmbedding, docEmbedding)

                    if similarity >= threshold {
                        var metadata: String? = nil
                        if sqlite3_column_type(stmt, 3) != SQLITE_NULL {
                            metadata = String(cString: sqlite3_column_text(stmt, 3))
                        }

                        let result = RetrievalResult(
                            id: id,
                            content: content,
                            similarity: similarity,
                            metadata: metadata
                        )
                        results.append((result: result, similarity: similarity))
                    }
                }
            }
        }

        sqlite3_finalize(stmt)

        // Sort by similarity (descending) and take top K
        return results
            .sorted { $0.similarity > $1.similarity }
            .prefix(topK)
            .map { $0.result }
    }

    /// Get vector store statistics
    func getStats() throws -> VectorStoreStats {
        guard let db = db else {
            throw VectorStoreError.databaseNotInitialized
        }

        let countSQL = "SELECT COUNT(*) FROM \(Self.tableName);"
        var stmt: OpaquePointer?
        var count: Int64 = 0

        if sqlite3_prepare_v2(db, countSQL, -1, &stmt, nil) == SQLITE_OK {
            if sqlite3_step(stmt) == SQLITE_ROW {
                count = sqlite3_column_int64(stmt, 0)
            }
        }

        sqlite3_finalize(stmt)

        return VectorStoreStats(
            documentCount: count,
            vectorDimension: Int64(detectedDimension ?? 0)
        )
    }

    /// Clear all documents from vector store
    func clear() throws {
        guard let db = db else {
            throw VectorStoreError.databaseNotInitialized
        }

        let deleteSQL = "DELETE FROM \(Self.tableName);"

        if sqlite3_exec(db, deleteSQL, nil, nil, nil) != SQLITE_OK {
            throw VectorStoreError.deleteFailed("Failed to clear vector store")
        }
    }

    /// Close database connection
    func close() {
        if let db = db {
            sqlite3_close(db)
            self.db = nil
        }
    }

    // MARK: - Private Methods

    private func createTable() throws {
        guard let db = db else {
            throw VectorStoreError.databaseNotInitialized
        }

        let createTableSQL = """
        CREATE TABLE IF NOT EXISTS \(Self.tableName) (
            \(Self.columnId) TEXT PRIMARY KEY,
            \(Self.columnContent) TEXT NOT NULL,
            \(Self.columnEmbedding) BLOB NOT NULL,
            \(Self.columnMetadata) TEXT,
            \(Self.columnCreatedAt) INTEGER DEFAULT (strftime('%s', 'now'))
        );
        CREATE INDEX IF NOT EXISTS idx_created_at ON \(Self.tableName)(\(Self.columnCreatedAt));
        """

        if sqlite3_exec(db, createTableSQL, nil, nil, nil) != SQLITE_OK {
            throw VectorStoreError.tableCreationFailed("Failed to create table")
        }
    }

    /// Convert embedding List<Double> to binary BLOB (float32)
    /// Format: Little-endian float32 array
    private func embeddingToBlob(_ embedding: [Double]) -> Data {
        var data = Data(count: embedding.count * 4)
        data.withUnsafeMutableBytes { ptr in
            let floatPtr = ptr.bindMemory(to: Float.self)
            for (i, value) in embedding.enumerated() {
                floatPtr[i] = Float(value)
            }
        }
        return data
    }

    /// Convert binary BLOB (float32) to embedding List<Double>
    private func blobToEmbedding(_ data: Data) -> [Double] {
        return data.withUnsafeBytes { ptr in
            let floatPtr = ptr.bindMemory(to: Float.self)
            return (0..<floatPtr.count).map { Double(floatPtr[$0]) }
        }
    }

    deinit {
        close()
    }
}

// MARK: - Error Types

enum VectorStoreError: Error, LocalizedError {
    case databaseNotInitialized
    case databaseOpenFailed(String)
    case tableCreationFailed(String)
    case insertFailed(String)
    case deleteFailed(String)
    case dimensionMismatch(expected: Int, actual: Int)

    var errorDescription: String? {
        switch self {
        case .databaseNotInitialized:
            return "Database not initialized. Call initialize() first."
        case .databaseOpenFailed(let message):
            return "Failed to open database: \(message)"
        case .tableCreationFailed(let message):
            return "Failed to create table: \(message)"
        case .insertFailed(let message):
            return "Failed to insert document: \(message)"
        case .deleteFailed(let message):
            return "Failed to delete: \(message)"
        case .dimensionMismatch(let expected, let actual):
            return "Embedding dimension mismatch: expected \(expected), got \(actual)"
        }
    }
}
